### ISSUE TITLE
Consider removing array sizes from their types.

### ISSUE THREAD

**Issue by otrho** [essential-contributions/pint#1017]
We currently store the size of an array in the type a couple of ways -- with a `range: Option<ExprKey>` and with `size: Option<i64>`.  Both are obviously optional and we may easily have an array type with unknown size.

Soon we'll add support explicitly for 'variable' or essentially unknown sized arrays for which the size is known only at runtime.  The size will be fixed and not actually variable, but only known at runtime.

Currently two arrays of different size have the same type, i.e., `int[2]` is the same as `int[222]`.

Having the size in the type is useful for bounds checking and confirming correct behaviours when using a valueless union (aka an enum) as the indexing range.  But it adds a lot of complexity to the type system -- it's the only reason `Type` has a dependency on `ExprKey` -- and it makes managing expressions much more complex.  Not only do we need to worry about exprs in `const` declarations, predicate parameters, `let` expressions and constraints, but also in any of their types, just in case they are arrays.  (But even then only optionally.)

I think that given we'll have explicitly unknown sized arrays coming soon and the above complexity, it makes sense to remove the array size from the type and for it to be stored only with the expression, if known.  This would help the type checker -- `const`s wouldn't need to be evaluated especially before type checking -- and would simplify the backend or other places where the array size is required.

The only question I have is perhaps whether to remove the size from the language.  Do we only allow an array type to be written as `int[]` or `bool[][][]` rather than having the explicit sizes?  We will need to allow `int[]` style declarations for unknown sized arrays when they arrive, so perhaps we have all arrays declared as such?

When using these types, they can be optionally attached to `const` and `let` decls, but they always have a RHS initialiser, which will dictate the array size.  `let ary: int[] = [1, 2, 3]` is explicitly a 3 element array, though its type doesn't know that.

Predicate parameters will need to allow unknown sized arrays anyway, so there is no initialiser and the length of the array is a runtime value.  A `len(ary)` intrinsic will be used.

**Comment by mohammadfawaz** [comment]
I like how much this simplifies things.

One thing we lose though is predicate parameters that are arrays of fixed size as in:
```rust
predicate foo(a: int[3]) {

}
```
where we would have to add constraint that guarantees that the size of `a` in the solution is 3. 

That being said, I think we could probably get around this by immediately converting this information as a constraint where the above would be transformed (by the compiler) to:
```rust
predicate foo(a: int[]) {
    constraint len(a) == 3;
}
```
So, the user would still encode size information between `[..]` but the compiler would move that information into a constraint. This can be done very early, even before type checking (or even by the parser like we used to do with initialized decision variables back in the days).

**Comment by otrho** [comment]
Hrmm, yep, it's tricky.  The main point here is to remove the optional range and size from the _array type_.  So having `int[3]` allowed sometimes means we'd have to store the size in the parameter itself somehow.  This might not be a bad idea in general, to always have the param sizes pre-computed/available.

**Comment by mohammadfawaz** [comment]
I was just thinking we would turn the size information into a constraint right away.. the `3` above wouldn't even go into the type.. it would just become a constraint in the parser and its purpose is constraining the solution. So the `3` will not actually be used for compile-time OOB checks for example.

**Comment by otrho** [comment]
> I was just thinking we would turn the size information into a constraint right away.. the `3` above wouldn't even go into the type.. it would just become a constraint in the parser and its purpose is constraining the solution. So the `3` will not actually be used for compile-time OOB checks for example.

Yikes, our parser already does waaay too much.  I guess it's already just pushing constraints into the `Predicate` as they're parsed, but if ever we re-do the parser/macro system in the future this wouldn't be possible.

**Comment by mohammadfawaz** [comment]
Yeah that's true. Maybe it's one of those things that has to get resolved right after the initial ast is produced (which will have the size in the array type) but the next phase doesn't have the size in the type (it would be a different `Type` enum)

**Comment by otrho** [comment]
While looking at implementing unsized arrays I've noted that sized arrays and unsized arrays need to remain distinct, mostly because unsized arrays cannot be put into other aggregate types, or probably themselves.

I.e., Sized arrays, tuples, unions require at compile time for their element sizes to be known.  Without this information constant O(1) access is not possible.  Iteration through arrays is an alternative, though a huge concession, but iteration for tuples and unions almost defeats the purpose with O(n) access for fields and the associated code bloat.

**Comment by mohammadfawaz** [comment]
For something like this:
```
type Tuple = {
    int[],
    b256[],
    bool[],
}
```
We know that the layout would look like:
```
[ n, i_1, ..., i_n, m, b_1, ..., b_m, w, c_1, .. c_w ]
```
where `n` is the length of the `int[]` array, `m` is the length of `b256[]` array, and `w` is the length of `bool[]` array. If I want to read `i_k`, the compiler has to compute this index:
```
1 + k
```
If I want to read `b_k`, the compiler has to compute this index:
```
n + 1 + k
```
If I want to read `c_k`, the compiler has to compute this index:
```
n + 1 + m + 1 + k
```
But of course it was to read `m` first from location `n + 1` so this becomes:
```
n + 1 + load(n + 1) + 1 + k
```
So yeah in the general case, we need to do multiple loads from memory before we can compute the final index.

In the case of a 2D array, then yeah we need a loop. Or the array has to be a matrix? So we don't allow this for exmaple
```
let a: int[][] = [ [ 1, 2 ], [ 3, 4, 5] ];
```
Instead, all the inner arrays must have the same size.

**Comment by otrho** [comment]
Getting `n` is also a load.   It depends if we have runtime OOB assertions too, which I think I _have_ started adding in my branch.  So even to read from the `int[]` it's a load of `n`, check `k < n`, then load from `1 + k`.  So even if the start of the array is at a known index it's two loads per read.  That's still O(1) though.  But for the others there's those extra reads of the prior arrays.

For nested arrays `int[][]` it becomes O(n) unless, as you say, the inner arrays are all the same length.  But there's no way to verify this, either at compile time or runtime.  For a statically typed language this is a pretty ugly corner, and it'd be a rule fairly easy to violate accidentally.

The O(n) iteration isn't the worst thing in the world if the code is iterating over the outer array anyway, with a morphism.  It's only a problem if the code is attempting arbitrary random access into the middle of it, repeatedly.  Documenting the difference, or even providing a linter (one day!) warning might be enough to alleviate the problem most of the time.

Would users truly expect random access into an `A[i][j]`?  I guess it depends on where they get `i` and `j` from.  If it's from a `map i in 0..__len(ary) { ... }` then the linter says 'no', use `map x in ary { .. }` or maybe `map i, x in ary { .. }` if we provide that.

Oh no, I've just considered `int[][][]`.  Getting `__len(..)` for the outermost array would be O(n) unless we continue to ban inner unsized arrays. :|
